transformer_engine/jax/praxis/transformer.py

```diff
@@ -0,0 +1,260 @@
+        """__call__"""
+        return self.relative_position_bias(q_seqlen, k_seqlen, bidirectional)
+
+
+class MultiHeadAttention(TransformerEngineBaseLayer):
+    """MultiHeadAttention"""
+
+    class HParams(TransformerEngineBaseLayer.HParams):    # pylint: disable=too-few-public-methods
+        """MultiHeadAttention.HParams"""
+        input_dims: int = 512
+        head_dim: int = 64
+        num_heads: int = 16
+        dropout_rate: float = 0.
+        dropout_rng_name: str = 'dropout'
+        layernorm_type: str = "layernorm"
+        layernorm_epsilon: float = 1e-6
+        use_bias: bool = False
+        bias_init: WeightInit = WeightInit.Constant(0.0)
+        apply_residual_connection_post_layernorm: bool = False
+        output_layernorm: bool = False
+        attn_type: AttentionType = AttentionType.PADDING
+        fuse_qkv: bool = True
+        transpose_batch_sequence: bool = True
+        scale_attn_logits: bool = False
+        scaled_query_init: bool = True
+        float32_logits: bool = False
+
+    def setup(self) -> None:
+        """setup"""
+        super().setup()
+
+        mha_cls = partial(flax_MultiHeadAttention,
+                          dtype=self.hparams.dtype,
+                          head_dim=self.hparams.head_dim,
+                          num_heads=self.hparams.num_heads,
+                          dropout_rate=self.hparams.dropout_rate,
+                          dropout_rng_name=self.hparams.dropout_rng_name,
+                          layernorm_type=self.hparams.layernorm_type,
+                          layernorm_epsilon=self.hparams.layernorm_epsilon,
+                          kernel_init=TransformerEngineBaseLayer.generate_params_init(
+                              "kernel", self.hparams.params_init),
+                          use_bias=self.hparams.use_bias,
+                          bias_init=TransformerEngineBaseLayer.generate_params_init(
+                              "bias", self.hparams.bias_init),
+                          apply_residual_connection_post_layernorm=self.hparams.
+                          apply_residual_connection_post_layernorm,
+                          output_layernorm=self.hparams.output_layernorm,
+                          attn_type=self.hparams.attn_type,
+                          fuse_qkv=self.hparams.fuse_qkv,
+                          transpose_batch_sequence=self.hparams.transpose_batch_sequence,
+                          scale_attn_logits=self.hparams.scale_attn_logits,
+                          scaled_query_init=self.hparams.scaled_query_init,
+                          float32_logits=self.hparams.float32_logits)
+
+        def module_init_args():
+            # (Batch, Seqlen, HIDDEN)
+            return (jnp.zeros((32, 128, self.hparams.input_dims), dtype=self.hparams.dtype),
+                    jnp.zeros((32, 128, self.hparams.input_dims), dtype=self.hparams.dtype))
```

Comment:
The shape here is only used to initialize weights, thus only `input_dims` is important.

Reply:
Makes sense. It's unintuitive though, so we should comment wherever we have these hard-coded dims. Also, could we set all the unused dims to 1? It would standardize things and dims like this seem like they're from real data.